runtimes/js/encore.dev/api/mod.ts

```diff
@@ -261,5 +265,128 @@ api.static = function staticAssets(options: StaticOptions) {
   return new StaticAssets(options);
 };
 
+export interface MiddlewareOptions {
+  /**
+   * Configuration for what endpoints that should be targeted by the middleware
+   */
+  target?: {
+    /**
+     * If set, only run middleware on endpoints that are either exposed or not
+     * exposed.
+     */
+    expose?: boolean;
+
+    /**
+     * If set, only run middleware on endpoints that either require or not
+     * requires auth.
+     */
+    auth?: boolean;
+
+    /**
+     * If set, only run middleware on endpoints that are raw endpoints.
+     */
+    isRaw?: boolean;
+
+    /**
+     * If set, only run middleware on endpoints that are stream endpoints.
+     */
+    isStream?: boolean;
+  };
+}
+
+export class MiddlewareRequest {
+  private _reqMeta?: RequestMeta;
+  private _stream?: Sink | Stream | Socket;
+  private _rawReq?: RawRequest;
+  private _rawResp?: RawResponse;
+
+  constructor(
+    stream?: Sink | Stream | Socket,
+    rawReq?: RawRequest,
+    rawResp?: RawResponse
+  ) {
+    this._stream = stream;
+    this._rawReq = rawReq;
+    this._rawResp = rawResp;
+  }
+
+  /**
+   * requestMeta is set when the handler is a typed handler or a stream handler.
+   * for raw handlers, see rawRequest and rawResponse.
+   */
+  public get requestMeta(): RequestMeta | undefined {
+    return this._reqMeta || (this._reqMeta = currentRequest());
+  }
+
+  /**
+   * stream is set when the handler is a stream handler.
+   */
+  public get stream(): Sink | Stream | Socket | undefined {
+    return this._stream;
+  }
+
+  /**
+   * rawRequest is set when the handler is a raw request handler.
+   */
+  public get rawRequest(): RawRequest | undefined {
+    return this._rawReq;
+  }
+
+  /**
+   * rawResponse is set when the handler is a raw request handler.
+   */
+  public get rawResponse(): RawResponse | undefined {
+    return this._rawResp;
+  }
+}
+
+export class HandlerResponse {
+  /**
+   * The payload returned by the handler when the handler is either
+   * a typed handler or stream handler.
+   */
+  payload: any;
+
+  /**
+   * extraHeaders can be used by middlewares to add headers to the
+   * response. This only works for typed handler. For raw handlers
+   * see MiddlewareRequest.rawResponse.
+   */
+  extraHeaders: Record<string, string | string[]>;
```

Comment:
I think API-wise it would be better to make this private and expose something like `addHeader` and `setHeader` or something to that effect. Rather than having the caller set the whole object. I think that would compose better with multiple middleware